### ISSUE TITLE
This diff contains a change in line endings from 'LF' to 'CRLF'.

### ISSUE THREAD

**Issue by aybe** [desktop/desktop#18225]
### The problem

This is shown for new files... quite confusing because there can be no changes in a new file.

![image](https://github.com/desktop/desktop/assets/1537591/bcc606a2-a664-4ce9-8e16-52c8b679f91a)

I believe it should only be shown on modified files, not new files.

### Release version

 3.3.9-beta2

### Operating system

Windows 11

**Comment by aybe** [comment]
Well, turns out that after testing, it is always shown and that is **really annoying**.

_This diff contains a change in line endings from 'LF' to 'CRLF'._
- LF to CRLF is an implementation detail IMO.

There should be an option to allow to turn it off because it's meaningless now as it's shown on every single file.

Hope you get the point :)

**Comment by alkorlos** [comment]
Today GitHub Desktop updated to 3.3.9 (x64), the same issue appeared.

Windows 10

**Comment by alkorlos** [comment]
@aybe 

On Windows, it's better not to use non standart lf EOL, exception only if you are sure that you need it. @LO-0116 warns about a [potential problem](https://github.com/desktop/desktop/issues/18225#issuecomment-2221889225).

This will set the default behavior:
`.gitattributes`
```
* text=auto
```

**Attention**
You would not see the warning `This diff contains a change in line endings from 'LF' to 'CRLF'.` if you were using standard EOL. Carefully consider which EOL are needed for which extensions to avoid problems.

[More details](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#example) about the options.

**Comment by aybe** [comment]
@alkorlos Oh yes, thanks :)

**Comment by tidy-dev** [comment]
👋  As you have found out, this warning has always been present. The new release makes it more prominent as part of accessibility efforts to ensure all users can access the warning. 

That said, I am going to label this as an enhancement request as we are discussing ideas to make it less annoying or at least provide information about why it is showing up and where to learn about the git configurations you may choose to set depending on your teams operating systems setup.

You can find more about modifying the git configs around `core.autocrlf` in the "Formatting  and Whitespaces" in the [Customizing Git - Git Configuration documentation](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration).  

[Configuring Git to handle line endings ](https://docs.github.com/en/get-started/getting-started-with-git/) also has more info on setting for this per repository and per operating system.

**Comment by aybe** [comment]
Ah yes, accessibility, makes sense now! I'll apply your suggestions, thanks everyone :)

**Comment by bert-w** [comment]
Might there also be an error in the text? I am getting the exact message

"This diff contains a change in line endings from 'LF' to 'CRLF'."

however my file is in LF format. The error message implies it is in CRLF format.

**Comment by twistios** [comment]
The same happens to me as it does for @bert-w. I have used the command `git config --global core.autocrlf true` (Windows system). The file actually is in LF-Format (according to Notepad++).

**Comment by alkorlos** [comment]
@bert-w This is correct text.

You, probably, changed the EOL to `lf` in the text editor or `.editorconfig`. Your file have `lf`.
But in windows defaulf file saves to git with `crlf`.
Messege about it — `This diff contains a change in line endings from 'LF' to 'CRLF'`

To fix this, you need to adjust the git settings in addition to changing the EOL in  text editor or `.editorconfig`. To save files in git with `lf` line endings:
`.gitattributes`
```
* text eol=lf
```
or
`git config core.autocrlf input`

The warning message is not well formulated. To understand it, one needs to know about the peculiarities of EOL and how git handles them by default.

P.S. On Windows, it's better not to use non standart `lf` EOL, except where they are necessary (for example, in `.sh` files).

**Comment by aybe** [comment]
Would like to add that the appearance of the message can be subtle depending the repository.

For instance, a Unity project in Windows:
 - you create scripts, they are saved as CRLF, it never shows up
 - you modify an asset that changes its meta, Unity saves it as LF, it shows up

So, the Unity repository actually use both CRLF and LF... pretty confusing!

**Comment by berniegp** [comment]
@aybe you might be interested in trying out this Unity `.gitattributes` template https://github.com/gitattributes/gitattributes/blob/master/Unity.gitattributes.

If you want you can then renormalize your line endings as described here https://stackoverflow.com/questions/15641259/how-to-normalize-working-tree-line-endings-in-git

**Comment by aybe** [comment]
@berniegp That worked, thanks!!! 👌🥳😁

**Comment by Kristyna131** [comment]
Duplicado de #

**Comment by kisontech** [comment]
i v many problems with wordpress with github

**Comment by webgeek-studio** [comment]
You can choose to use 'CRLF', 'LF', or even 'auto', which allows Git to automatically determine and use the appropriate line endings format based on your operating system.

To set the line endings format in your repository, you can use the git config command. For example, to set the line endings format to 'LF', you can execute the following command:

`git config --global core.autocrlf input`

This configures Git to automatically convert line endings to 'LF' upon committing changes. If you're working on Windows and want to keep 'CRLF', you can use:

`git config --global core.autocrlf true`

After applying the settings, modify the files in your project and commit the changes. Now, GitHub Desktop should not show you a warning about changing the line endings format.

**Comment by Genius-pig** [comment]
if you develop a unreal project, use `* text eol=crlf` in your `.gitattributes`.

**Comment by rsb-23** [comment]
In my case, even though I have modified `from CRLF to LF`; it gives same error `from LF to CRLF`.
It seems this case isn't handled, hence its a bug.

I verified the case by adding pre-commit hook to check for CRLF and it couldn't find it.

**Comment by AmodeusR** [comment]
> To set the line endings format in your repository, you can use the git config command. For example, to set the line endings format to 'LF', you can execute the following command:
> 
> `git config --global core.autocrlf input`

Thanks, this worked for me, but I have a question about it: I will need to do it in every repo I do from now on? Isn't there a way it can be configured globally to apply to every project?

**Comment by aybe** [comment]
@berniegp @webgeek-studio 

I thought things were solved until I upgraded Unity version:

<img width="465" alt="image" src="https://github.com/desktop/desktop/assets/1537591/d3d825f3-2317-46cc-9184-2450b6232365">

Tried `git config --global core.autocrlf true` and it finally worked. 🥳

Let's see how it holds up over the next weeks!

**Comment by aybe** [comment]
I talked too fast... `git config --global core.autocrlf true` fixes text files but screws scripts...

Falling back to `.gitattributes` for now.

**Comment by berniegp** [comment]
@aybe Indeed looking at the [.gitattributes](https://github.com/gitattributes/gitattributes/blob/master/Unity.gitattributes) template I see that `.txt` files aren't covered. It will probably work if you add this line:
```
*.txt                   text
```

Or maybe this one to affect only `ProjectSettings/ProjectVersion.txt`:
```
ProjectSettings/ProjectVersion.txt text
```

**Comment by AlisterH** [comment]
> Might there also be an error in the text? I am getting the exact message
> 
> "This diff contains a change in line endings from 'LF' to 'CRLF'."
> 
> however my file is in LF format. The error message implies it is in CRLF format.

Yes.  Here I have this message with a file that has changed from CRLF to LF.  So the message is backwards.

**Comment by LO-0116** [comment]
WARNING TO ANYONE ADDING `* text eol=lf` IN `.gitattributes`:
This will cause .png files to be uploaded incorrectly, so that the ones in your repository will become unusable. This cost me months of work and nobody has said anything about it anywhere. I'm not certain if other file formats will be affected.

**Comment by joebowbeer** [comment]
I'm seeing this when I created a new repo on Windows and then ran `cargo init`

I noticed that `gh` had added a `.gitattributes` with `text=auto` even though I explicitly configured `autocrlf=false` in my global git config.

Now, after running `cargo init`, `gh` is informing me that it changed (or will change?) all the line endings in the files added by cargo from LF to CRLF.

So I'm looking for a setting (other than git config?!) to configure this initial `.gitattributes`  but I'm not finding it.

Is my analysis correct?

Should I create a new issue for disabling `.gitattributes`?

Workaround: Remember to remove `.gitattributes`

**Comment by bmitc** [comment]
> As you have found out, this warning has always been present. The new release makes it more prominent as part of accessibility efforts to ensure all users can access the warning.

@tidy-dev Is this true? I have not seen this warning in a long time, which is why I'm here. (Although granted, I often use VS Code's source-code control now.) This is because I always configure Git to leave line-endings alone. It doesn't belong in Git, and all modern editors handle this. So I configure Git with `git config --global core.autocrlf false` and then configure Visual Studio Code to always use `LF`.

----

Here is the message I get for a Git configuration with `git config --global core.autocrlf false` and the file already using `LF` when committing *new* files:

![image](https://github.com/user-attachments/assets/92c7b9fc-8fcb-45db-92b4-e946197e4017)

I strongly disagree that addressing this is a "feature enhancement" because there are only two options here:

1. The message is "correct" in that GitHub Desktop is going to change my `LF` into `CRLF` upon commit, which completely disregards my global Git configuration. That is a bug.

2. The message otherwise makes no sense, especially since these are new files, and there's a no operation. That is a bug.

Either way, I have no idea what's going to happen when I commit here if using GitHub Desktop. I don't even understand the message itself to begin with.

**Comment by bmitc** [comment]
I apologize for an immediate follow-up, but it just occurred to me what's going on. I just happened to use GitHub Desktop to create the repository for me. In doing so, it automatically created a `.gitattributes` file that I didn't notice with the content:

```
# Auto detect text files and perform LF normalization
* text=auto
```

This should be considered incorrect behavior in that GitHub Desktop should pay attention to the global Git configuration settings instead of automatically generating the `.gitattributes` with instructions that are the complete opposite of the global configuration.

So I think there are two follow-ups here:

1. It would be nice if the message said where the configuration to do the line-ending conversion is coming from.
2. GitHub Desktop should honor global Git configurations and not automatically generate `.gitattributes` settings based only upon the OS.

**Comment by joebowbeer** [comment]
@bmitc I concur. This is a misfeature or bug in GH Desktop.

https://github.com/desktop/desktop/issues/18225#issuecomment-2294633122

**Comment by TonyWu-21474** [comment]
same issue when using Github Desktop uploading a MDK-ARM project (*.uvprojx )
![image](https://github.com/user-attachments/assets/4d4e2556-c542-4f6c-b3c0-851812124d37)

**Comment by jontuo** [comment]
I was going to try Github Desktop. It wants me to commit changes to 1176 files because of "This diff contains a change in line endings from 'LF' to 'CRLF'.". Ive been googling and loads of different answers.

**Comment by joebowbeer** [comment]
@jontuo Workaround: Remember to remove .gitattributes

**Comment by foxt** [comment]
Can anyone actually explain what this message means? I'm working on a project where the files are stored in WSL and then the repo is opened in Desktop (native on Windows, I don't think Desktop is supported on Ubuntu anyway)

To me, the message "This diff contains a change in line endings from 'LF' to 'CRLF'." reads to me that the file was originally saved as LF, but was converted to CRLF on disk - which isn't true, my VSCode is set to save as LF, not CRLF, and opening the file in NP++ in Windows confirms it's LF. The message doesn't say anything about if the file will be auto-converted on commit

**Comment by jimcullenaus** [comment]
@alkorlos 

> You, probably, changed the EOL to `lf` in the text editor or `.editorconfig`. Your file have `lf`.

You are correct, in my case.

> To fix this, you need to adjust the git settings in addition to changing the EOL in text editor or `.editorconfig`. To save files in git with `lf` line endings: `.gitattributes`
> 
> ```
> * text eol=lf
> ```
> 
> or `git config core.autocrlf input`

But I already have. If I open PowerShell, navigate to the directory of my repo, and type `git config core.autocrlf`, the output is `input`. And yet GitHub Desktop is still telling me "This diff contains a change in line endings from 'LF' to 'CRLF'."

I have no `.gitattributes` file.

**Comment by alkorlos** [comment]
@jimcullenaus

Rechecked how `git config core.autocrlf input` works in a new project:

1. Created a repository
2. Created a `.editorconfig` with:
```
root = true

[*]  
end_of_line = lf  
```
3. Created a file, ensured it had LF line endings, and confirmed that GitHub Desktop showed a warning
4. In the repository folder ran `git config core.autocrlf input`, and the warning in GitHub Desktop disappeared

I can't say why it didn’t work for you. If `git config core.autocrlf` returns `input`, there should be no warning for files with LF line endings.

**Comment by joebowbeer** [comment]
@alkorlos to be clear, are you claiming that the problem reported by OP is not a bug or misfeature?

**Comment by alkorlos** [comment]
@joebowbeer The warning message displayed by GitHub Desktop is correct, but it not well formulated. To understand it, one needs to know about the peculiarities of EOL and how Git handles them by default.

The contributor mentioned above that they want to improve the clarity of the warning, which would indeed be helpful.

**Comment by joebowbeer** [comment]
@alkorlos I think you are mistaken. This is a misfeature of the desktop app.

See my previous comments.

It looks to me like the desktop app is maybe sort-of trying to be helpful in one common use case, but in the process is it creating extra work for other configurations.

There is no way that documentation can help here. There should be no need for desktop to create this situation in the first place.

**Comment by alkorlos** [comment]
@joebowbeer 

> This is a misfeature of the desktop app.

Are you referring to GitHub Desktop creating `.gitattributes` by default? I think this goes beyond the scope of this issue. If you think the behavior regarding `.gitattributes` should be changed, it might be better to create a separate issue.

This issue is about the warning: `This diff contains a change in line endings from 'LF' to 'CRLF'`. I was only referring to that above.

This warning can also appear when creating a repository in other ways. If, on Windows, `.editorconfig` specifies LF line endings for files and there are no corresponding global or local Git settings, this warning will appear. This behavior is correct, although it could be improved.

**Comment by Andrej730** [comment]
Warning message `This diff contains a change in line endings from 'LF' to 'CRLF'.` is indeed misleading, for the reader it seems to imply that file was LF on index and user is trying to commit CRLF, which would imply that line ending type for this file will be changed on index too.

I don't know if current message is entirely technically correct - though it does communicate that something happened to line endings, it leaves user confused.
It's seems it should be `This diff contains a change in line endings from 'CRLF' to 'LF'.`, since this is exactly what happens - user changed EOL from CRLF to LF and therefore warning should state EOL was changed from CRLF to LF.

**Comment by Bakr-Ali** [comment]
@alkorlos The correct gitattribute setting is `* text=auto eol=lf` not `* text eol=lf` and there is a big difference between them.
You should _**never**_ use `* text eol=lf`! It means to set **_all_** files `*` as text `text` and then make their line endings LF `eol=lf`

That's why @LO-0116 was having the problem with PNGs.

See for explanation: https://github.com/desktop/desktop/issues/5323#issuecomment-410220449
> `* text eol=lf`
> This is the problem line - this rule essentially says "for all files attempt to do line ending conversion and ensure the line endings are set to LF". That's why Git has gone and changed the line endings inside the [PNG file signature](http://www.libpng.org/pub/png/spec/1.2/PNG-Rationale.html#R.PNG-file-signature) - it's spotted the CRLF at the start of the file and applied this filter rule to it.

Or just Google `* text` vs `* text=auto` or ask a LLM for the difference.

So please update your comments above accordingly as yours is the most upvoted and so people will use it and encounter issues.

**Comment by joebowbeer** [comment]
@alkorlos wrote:

> > This is a misfeature of the desktop app.
> 
> Are you referring to GitHub Desktop creating `.gitattributes` by default? I think this goes beyond the scope of this issue. If you think the behavior regarding `.gitattributes` should be changed, it might be better to create a separate issue.

Thanks for clarifying.

I asked about creating a new issue in my initial comment: https://github.com/desktop/desktop/issues/18225#issuecomment-2294633122

The next time I need to remove a `.gitattributes` that GitHub Desktop foisted on me, I'll create a new issue and reference this.

**Comment by alkorlos** [comment]
@Bakr-Ali Yes, you are right, `* text eol=lf` should not be used as a general recommendation. I added a warning about this to the comment a long time ago, but I intentionally did not remove the older parts in order to preserve the history of the discussion.

Looked at it again and agree: in its previous form, the comment could be misleading. Removed the old part.
